### DISCUSSION TITLE
Question about VAMPIRE2

### DISCUSSION THREAD

**Discussion by insujeon** [cnguyen10/few_shot_meta_learning#21] [Q&A] [answered]
Hi! I have some questions about the implementation of VAMPIRE2.

1) In line 96 of your implementation of Vampire2.py, the (adapted) variational parameters (mu_q and log_std_q, or fast_params) are also trained with the KL divergence on the validation set. Could you explain why there should be a KL loss term since line 9 of Algorithm 1 in the paper only show a loglikelihood term? (is this for backpropagating the gradient to the theta?)

2) Can you explain why the KL term is defined on Normal distribution? I might understand this as since we initialize the BNN parameter theta with Normal distribution; the prior is defined as Normal. But shouldn't we update theta with the validation set? Then, shouldn't we use the new prior distribution (with updated theta) as a base point in the KL term (similar to line 40 in Abml.py)?

By the way, I really appreciate all your work, including the paper and implementation.
It really broadens my knowledge and helps my current study on Bayesian meta-learning.
I am looking forward to discussing the above two questions (or other Bayesian meta-learning stuff) with you. 
Since there is no one around me to discuss this subject. 
Thank you, and have a nice day!

**Comment by cnguyen10** [comment, marked as answer]
Thank you for your interest in the implementation, Insu Jeon.

1. For your first concern, yes, it is intended in version 2. In version 1, only the loss on the training subset had the KL divergence, while the one on the validation subset did not. However, as I realised later, meta-learning was an extended version of hyper-parameter optimisation. In that case, we wanted to learn a good initialisation in terms of distributions. Thus, both the losses in training and validation should be consistent. That is why I added the KL divergence in the loss evaluated on the validation subset. As I tested, there was not much difference in terms of results.
2. As explained in our paper, our method was a meta-learning version of *Bayes by backprop*. We can select any distribution family as the variational posterior of $\theta$. For simplicity, we selected the multivariate normal distribution. For the reason why the prior is still a standard normal distribution, as I mentioned when answering your question 1, we want the losses in the training and validation consistent. In version 1, I think that I did something similar to ABML method, although we were not aware of it until one reviewer pointed out to us. In my opinion, you can also do that as well.
Let me know if you have any questions regarding to the implementation.
By the way, PyTorch now supports the functional form of neural networks under *functorch* library. You might re-implement it with that new library rather than the *higher* library.

**Reply by insujeon** [reply]
Thank you for your quick reply!
